tests/kernels/test_self_and_cross_attn.py

```diff
@@ -0,0 +1,1711 @@
+    * context_lens: list of context length values for each seq
+    * device: CPU or CUDA device
+
+    Returns:
+
+    * seq_lens_tensor: seq_lens list, as tensor
+    * context_lens_tensor: context_lens list, as tensor
+    * max_query_len: max(seq_lens) if is_seq, o/w 1
+    * max_context_len: max(context_lens)
+    * max_seq_len: max(seq_lens)
+    * seq_start_loc: start idx of each sequence
+    * query_start_loc: start idx of each query
+    '''
+    seq_lens_tensor = torch.tensor(seq_lens, dtype=torch.int, device=device)
+    context_lens_tensor = None if context_lens is None else torch.tensor(
+        context_lens, dtype=torch.int, device=device)
+    max_context_len = None if context_lens is None else max(context_lens)
+    max_seq_len = None if seq_lens is None else max(seq_lens)
+
+    seq_start_loc = torch.zeros(seq_lens_tensor.shape[0] + 1,
+                                dtype=torch.int32,
+                                device=device)
+
+    torch.cumsum(seq_lens_tensor,
+                 dim=0,
+                 dtype=seq_start_loc.dtype,
+                 out=seq_start_loc[1:])
+
+    if is_prompt:
+        # Prefill: query_start_loc matches seq_start_loc
+        query_start_loc = copy.deepcopy(seq_start_loc)
+        max_query_len = max_seq_len
+    else:
+        # Decode: one new query input token per batch element, thus
+        # query_start_loc is the cumsum of [1,1,1,...]
+        query_start_loc = list(range(len(seq_start_loc)))
+        max_query_len = 1
+
+    return seq_lens_tensor, \
+           context_lens_tensor, \
+           max_query_len, \
+           max_context_len, \
+           max_seq_len, \
+           seq_start_loc, \
+           query_start_loc
+
+
+def make_kv_cache(num_blocks: int,
+                  num_heads: int,
+                  head_size: int,
+                  block_size: int,
+                  device: Union[torch.device, str] = \
+                    CUDA_DEVICE,
+                  default_val: float=0.0) -> torch.Tensor:
+    '''
+    Create a fake KV cache.
+
+    Arguments:
+
+    * num_blocks: number of blocks in the KV cache
+    * num_heads: number of attention heads
+    * head_size: head dimension
+    * block_size: number of offsets within a block
+    * device: CPU or CUDA device
+    * default_val: initialization value for KV cache elements
+
+    Returns:
+
+    * kv_cache: 2 x num_blocks x (block_size * num_heads * head_size)
+    '''
+
+    kv_cache = torch.rand(
+        (2, num_blocks, block_size * num_heads * head_size)).to(device)
+    if default_val is not None:
+        kv_cache[:, :, :] = default_val
+    return kv_cache
+
+
+def num_tokens_to_min_blocks(num_tokens: int, block_size: int) -> int:
+    '''
+    Compute the minimum number of blocks required to hold num_tokens tokens,
+    given block_size
+    '''
+    return (num_tokens + block_size) // block_size
+
+
+def make_block_tables_slot_mapping(block_size: int,
```

Comment:
can we have this method return the block_table and slot_mappings for the complete sequences and in the caller use this return value to construct the needed block table tensors and mappings as needed. The reason I am suggesting this is that in function enc_dec_cross_attn_setup_reuses_query when we are invoking this method we are assigning slot_mapping to decode_slot_mapping (instead of using the decode_slot_mapping) which is confusing.

Reply:
Thanks! Makes sense. I refactored slot-mapping and block-table construction into multiple steps. 

make_empty_block_tables_tensor() builds an empty block-tables tensor (this models the behavior of ModelRunner in prefill phase)

make_empty_slot_mapping_tensor() builds an empty slot-mapping tensor (this is appropriate for enc/dec cross-attention *decode-phase*, since KV sequence length is static so newly decoded tokens do not require growing the KV cache)

make_block_tables_slot_mapping() returns
* block_tables_tensor # block tables for the full sequences
* slot_mapping_list # slot mapping for the full sequences
* max_block_idx # max address in the block table

split_slot_mapping() splits the full slot_mapping_list into prefill and decode slot mappings 

I added the following explanatory comments:

```
# Build prefill- & decode-phase data structures
# for decoder self-attention. Block tables and
# slot mapping must be in a format compatible
# with KV caching & attention kernels
#
# Prefill-phase:
#
# * Empty block-tables tensor
# * Slot-mapping with entries for prompt tokens
#
# Decode-phase:
# * Block-tables tensor with minimum number of blocks
#   required by total num. tokens in the entirety of all sequences
#   (including both prefill & decode)
# * Slot-mapping with entries for tokens that will be decoded in the
#   current decode iteration

...

# Build prefill- & decode-phase data structures
# for encoder/decoder cross-attention. Block tables and
# slot mapping must be in a format compatible
# with KV caching & attention kernels
#
# Whereas decoder self-attention extracts relationships between
# equal-length Q/K/V sequences, which mutually grow in length
# with each decoded token, cross-attention relates the Q sequence
# - which grows with each new decoded token - to fixed-length
# K and V sequences derived from the encoder hidden states.
#
# Prefill-phase:
#
# * Empty block-tables tensor
# * Slot-mapping with as many entries as there are tokens in the encoder
#   prompt.
#
# Decode-phase:
# * Block-tables tensor with minimum number of blocks to
#   accommodate K & V tensors which are equal in lnegth
#   to the encoder prompt length
# * Empty slot-mapping tensor (since K & V are fixed in size,
#   new decoded tokens are not KV-cached and require no slot-
#   mapping)
```